Firestore/Protos/nanopb/firestore/local/maybe_document.nanopb.cc

```diff
@@ -74,6 +76,57 @@ PB_STATIC_ASSERT((pb_membersize(firestore_client_NoDocument, read_time) < 256 &&
 #endif
 
 
+std::string firestore_client_NoDocument::ToString(int indent) const {
+    std::string header = PrintHeader(indent, "NoDocument", this);
+    std::string result;
+
+    result += PrintPrimitiveField("name: ", name, indent + 1, false);
+    result += PrintMessageField("read_time ", read_time, indent + 1, false);
+
+    std::string tail = PrintTail(indent);
+    return header + result + tail;
+}
+
+std::string firestore_client_UnknownDocument::ToString(int indent) const {
+    std::string header = PrintHeader(indent, "UnknownDocument", this);
+    std::string result;
+
+    result += PrintPrimitiveField("name: ", name, indent + 1, false);
+    result += PrintMessageField("version ", version, indent + 1, false);
+
+    std::string tail = PrintTail(indent);
+    return header + result + tail;
+}
+
+std::string firestore_client_MaybeDocument::ToString(int indent) const {
+    std::string header = PrintHeader(indent, "MaybeDocument", this);
+    std::string result;
+
+    switch (which_document_type) {
+    case firestore_client_MaybeDocument_no_document_tag:
+        result += PrintMessageField("no_document ",
+            no_document, indent + 1, true);
+        break;
+    case firestore_client_MaybeDocument_document_tag:
+        result += PrintMessageField("document ", document, indent + 1, true);
+        break;
+    case firestore_client_MaybeDocument_unknown_document_tag:
+        result += PrintMessageField("unknown_document ",
+            unknown_document, indent + 1, true);
+        break;
+    }
+    result += PrintPrimitiveField("has_committed_mutations: ",
+        has_committed_mutations, indent + 1, false);
+
+    bool is_root = indent == 0;
+    if (!result.empty() || is_root) {
```

Comment:
This is admittedly not the clearest way (because we generate `has_` fields very sparingly), but it should achieve the right outcome (modulo any bugs).

By default, all submessages are unset. The only way we have of setting a submessage without modifying any of its members is by setting the `has_` flag to `true`. In the current implementation, an unset submessage and an empty-but-set submessage are printed differently (the former is not printed, and the latter is printed as empty braces). If we were to always print submessages, we would lose that distinction (the submessages would always be considered set). This would be inconsistent with the (presumably conformant) output that Objective-C library produces, and would also mean that the generated text representation would not roundtrip if it were given as input to a text proto parsing function.

> This means that in our logs a submessage might be omitted even though we're actually sending it.

Unless there's a bug, this shouldn't happen in the current implementation. For an "optional" field (the one that has a dedicated `has_` flag), the generated code will always print the value if it's set, even if the value is  the default one (via the `always_print` parameter).

> If the submessage is not optional, always print its contents, even if all the fields have default values.

This is the case in the current implementation, though the condition is in `PrintMessageField`, not `ToString`.

I've added a unit test to show an empty optional is printed, but the larger point is that for our integration test suite, the generated output is _exactly_ the same as what the Objective-C library would generate (except for 2-vs-4 space indentation of the root message and the "headers" (e.g., `<Write 0x123>: {` vs `<GCFSWrite 0x456>: {`). It is possible that there are cases not triggered by out integration tests, but it's very unlikely that the current implementation is simply broken.

Reply:
The difference I'm trying to highlight is that fidelity to the Objective-C output is a non-goal. Fidelity to what we're actually sending is *the* feature we should be striving for here.

Objective-C protobuf treats all submessages as optional. nanopb treats them as optional only if we specially mark them up (which we've done only minimally). Because we're so conservative with making submessages optional (and minimally having `has_` accessors) we're potentially sending more submessages without meaningful content to the server. When the backend team files a bug against us I want the logs to reflect what we actually sent rather than what we might have sent if we were still using Objective-C protobuf.

Reply:
Discussed offline: the last round of comments does not apply because I misunderstood the nature of nanopb's default proto3 support. By default nanopb elides submessages whose values are all default-values. Flagging these as proto2 and giving them explicit `has_` accessors allows us to consider a message set even though it might otherwise have been elided from the output.

This means that this code that emulates that behavior is correct.

Reply:
Thanks for writing this up. I'll also copy the results of my experiments below, in case they come in handy later.

Imagine there's an interactive prompt that outputs the serialized form of the message after each operation, using `ByteString::ToString()` and also surrounding the output with quotes.

First, here's the current form of `StructuredQuery`, which treats `start_at` field (of type `Cursor`) as a proto3 field (that is, it doesn't have a `has_` flag for the field):
```
> Message<google_firestore_v1_StructuredQuery> m;
""
> m->start_at.before = true;
":\002\020\001"
```

Now, if I set options for that field to consider it a proto2 field, it would have a `has_start_at` flag generated for it, and the output would look like this:
```
> Message<google_firestore_v1_StructuredQuery> m;
""
> m->has_start_at = true;
":\000"
> m->start_at.before = true;
":\002\020\001"
> m->has_start_at = false;
""
```